src/qibolab/instruments/qm.py

```diff
@@ -676,31 +724,34 @@ class QMOPX(Controller):
     Args:
         name (str): Name of the instrument instance.
         address (str): IP address and port for connecting to the OPX instruments.
-
-    Attributes:
-        is_connected (bool): Boolean that shows whether instruments are connected.
-        manager (:class:`qm.QuantumMachinesManager.QuantumMachinesManager`): Manager object
-            used for controlling the QM OPXs.
-        config (dict): Configuration dictionary required for pulse execution on the OPXs.
-        time_of_flight (int): Time of flight used for hardware signal integration.
-        smearing (int): Smearing used for hardware signal integration.
     """
 
-    def __init__(self, name, address):
-        # QuantumMachines manager is instantiated in ``platform.connect``
-        self.name = name
-        self.address = address
-        self.manager = None
-        self.is_connected = False
-
-        self.time_of_flight = 0
-        self.smearing = 0
-        # copied from qblox runcard, not used here yet
-        # hardware_avg: 1024
-        # sampling_rate: 1_000_000_000
-        # repetition_duration: 200_000
-        # minimum_delay_between_instructions: 4
-        self.config = QMConfig()
+    name: str
+    address: str
+
+    manager: Optional[QuantumMachinesManager] = None
+    """Manager object used for controlling the QM OPXs."""
+    config: QMConfig = field(default_factory=QMConfig)
+    """Configuration dictionary required for pulse execution on the OPXs."""
+    is_connected: bool = False
+    """Boolean that shows whether we are connected to the QM manager."""
+    time_of_flight: int = 0
+    """Time of flight used for hardware signal integration."""
+    smearing: int = 0
+    """Smearing used for hardware signal integration."""
+    ports: Dict[Tuple[PortType], QMPort] = field(default_factory=dict)
+    """Dictionary holding the ports of controllers that are connected."""
+
+    def __getitem__(self, value):
```

Comment:
I renamed this to `port_name` that I believe better describes what the variable is at this context.

I am thinking to rename (not use) `__getitem__` for this purpose (getting the ports) to something like `get_port`. This is because this will be part of the driver (`Instrument`) interface so it should be clear to everyone what it does. The reason I am not sure if I should do this is because using the `__getitem__` simplifies a lot the platform creation methods:
```py
channels |= Channel("name", port=instrument["myportname"])
```
vs
```py
channels |= Channel("name", port=instrument.get_port("myportname"))
```

Reply:
I like `__getitem__` as well, well, and I see nothing bad in using it.
It just suggests the idea of `instrument` to be a `Port`s container. It is not fully matching, but it is not even far from truth (it's just *not only* a ports container).

In case you still want a named function, just use `.port()`. The initial `get_` adds nothing (as usual).